File: erlang/etudes/ch4/geom.erl
%% @author Howard Fear 
%% @doc Functions calculating various geometric properties

-module(geom).
-export([area/1]).

%% @doc Calculates area of various shapes.

-spec area({atom(), number(), number()}) -> number().

area({Shape, A, B}) -> area(Shape, A, B).
area(Shape, A, B) ->
    case Shape of
        rectangle when A > 0, B > 0 -> A * B;
        triangle when A > 0, B > 0 -> (A * B) / 2;
        ellipse when A > 0, B > 0 -> math:pi() * A * B;
        _ -> 0
    end.

File: elixir/etudes/ch04/geom.ex
defmodule Geom do
  def area(shape, a, b) when a >= 0 and b >= 0 do
    case shape do
      :rectangle -> a * b
      :triangle -> (a * b) / 2.0
      :ellipse -> :math.pi() * a * b
    end
  end
  def area(_, _, _) do
    0
  end
end
